### ISSUE TITLE
API to get snippets (handling of multiple manager instances)

### ISSUE THREAD

**Issue by blueyed** [SirVer/ultisnips#521]
I've noticed that the UltiSnips integration in YouCompleteMe is broken, when using `g:UltiSnipsUsePythonVersion = 3`.

This is because YouCompleteMe still requires and uses Python 2, and will then import another `UltiSnips_Manager` instance.

While the workaround is to use `g:UltiSnipsUsePythonVersion = 2` instead, I'd like to avoid this (because I use Python 3 by default, and want to avoid loading two interpreters).

YCM uses this method:

```
def _AddUltiSnipsDataIfNeeded( extra_data ):
  if not USE_ULTISNIPS_DATA:
    return

  try:
    rawsnips = UltiSnips_Manager._snips( '', 1 )
  except Exception as e:
    return

  # UltiSnips_Manager._snips() returns a class instance where:
  # class.trigger - name of snippet trigger word ( e.g. defn or testcase )
  # class.description - description of the snippet
  extra_data[ 'ultisnips_snippets' ] = [ { 'trigger': x.trigger,
                                           'description': x.description
                                         } for x in rawsnips ]
```

A fix might be to explicitly call the following before calling `_snips()`:

```
if vim.current.buffer.number not in UltiSnips_Manager._buffer_filetypes:
    UltiSnips_Manager.add_buffer_filetypes(vim.eval('&ft'))
```

That's what UltiSnips does itself on filetype changes:

```
function! UltiSnips#FileTypeChanged()
    echom "UltiSnips#FileTypeChanged"
    exec g:_uspy "UltiSnips_Manager.reset_buffer_filetypes()"
    exec g:_uspy "UltiSnips_Manager.add_buffer_filetypes('" . &ft . "')"
    return ""
endfunction
```

I think UltiSnips should provide an API for this, which YCM could use then.

(I've requested this in #315 before, although the discussion was mainly about the caching aspect then).

**Comment by SirVer** [comment]
I agree, an API would be very nice to have. In fact UltiSnips should be a Library with a tin client that uses that library - unfortunately I did not foresee that it would become so popular when I started it. 

I am very thin on time these days, so I think it is unlikely that I will work on such a fundamental refactor any time soon - if at all. I am totally willing to provide suggestions and guidance for an implementor of course.
